sources/osg/TimerGPU.js

```diff
@@ -0,0 +1,273 @@
+'use strict';
+
+// use EXT_disjoint_timer_query
+// to time webgl calls GPU side
+// average over multiple frames
+// for consistent results
+// use double buffer queries for that.
+// see http://www.reedbeta.com/blog/2011/10/12/gpu-profiling-101/
+var TimerGPU = function ( gl ) {
+
+    this._enabled = false;
+
+    if ( gl ) {
+
+        var ext = gl.getExtension( 'EXT_disjoint_timer_query' );
+        if ( !ext ) return;
+        // https://github.com/KhronosGroup/WebGL/blob/master/sdk/tests/conformance/extensions/ext-disjoint-timer-query.html#L102
+        this._hasTimeElapsed = ext.getQueryEXT( ext.TIME_ELAPSED_EXT, ext.QUERY_COUNTER_BITS_EXT ) >= 30;
+        this._hasTimeStamp = ext.getQueryEXT( ext.TIMESTAMP_EXT, ext.QUERY_COUNTER_BITS_EXT ) >= 30;
+
+        if ( !this._hasTimeElapsed && !this.hasTimeStamp ) {
+            return;
+        }
+
+        // no timestamp means not start/end absolute time
+        // which means each start must be followed by a end
+        // BEFORE any other start (of other queryID)
+        if ( !this.hasTimeStamp ) {
+            console.log( 'Warning: do not use interleaved GPU query' );
+        }
+
+        this._gl = gl;
+        this._glTimer = ext;
+        this._enabled = true;
+
+    }
+
+    // those we seek results of
+    // (all query per frame)
+    // double Buffered
+    this._pollingStartQueries = {};
+    this._pollingEndQueries = {};
+    // query list currently recording
+    // between a start and a end query
+    this._runningQueries = {};
+    // number of query asked (current index query per queryID)
+    this._timingCountQuery = {};
+    // number of query answered with results
+    this._resultCountQuery = {};
+    // cumulative average
+    this._averageTimerQuery = {};
+    // query waiting async results from GPU 
+    this._waitingQueries = {};
+};
+
+TimerGPU.instance = function ( gl ) {
+
+    if ( !TimerGPU._instance ) {
+        TimerGPU._instance = new TimerGPU( gl );
+    } else if ( gl && TimerGPU._instance.getContext() !== gl ) {
+        TimerGPU._instance.setContext( gl );
```

Comment:
In case first call to timerGPU.instance() is without the gl parameter. 
Cannot remember how that did happen.
Will remove.